### ISSUE TITLE
Concatenating files can generate same hash?

### ISSUE THREAD

**Issue by richardhuaaa** [mac-/hash-files#4]
As far as I understand, this works by concatenating all of the files together and then hashing the result. Does this mean that we can get the same hash when hashing this:

File 1: 00 01 02
File 2: 03 04 05

As well as this?

File 1: 00 01 02 03
File 2: 04 05

**Comment by mac-** [comment]
That's true, thanks for uncovering this issue. I'll get this fixed soon.

**Comment by richardhuaaa** [comment]
Oh, just another thing to note, I'm not sure how important this is to you, but this hashing algorithm also doesn't seem to factor in the names of the files. So if a file got renamed from `a.js` to `aa.js`, the hash would still come out the same.

**Comment by monteiz** [comment]
To solve both the issues at once, it would be enough to concatenate the full path of the file to the file data, and then hash.

I would create a PR, but I suppose this project is abandoned.

**Comment by andersk** [comment]
> To solve both the issues at once, it would be enough to concatenate the full path of the file to the file data, and then hash.

@monteiz No, because then you’d still get collisions like this:

- file `file1.txt` with contents `_foo_file2.txt_bar_` and file `file2.txt` with contents `_baz_`
- file `file1.txt` with contents `_foo_` and file `file2.txt` with contents `_bar_file2.txt_baz_`

which would both be smashed into `file1.txt_foo_file2.txt_bar_file2.txt_baz_` and hashed to the same result.

To solve these issues, you need to encode the filenames and corresponding data into a stream from which they could be *unambiguously recovered*, for example, `files.map((file) => file.name.length + " " + file.name + file.data.length + " " + file.data).join("")`.

**Comment by monteiz** [comment]
@andersk good observation
